File: backend/app/schemas/tax.py
"""Tax engine Pydantic schemas for API responses."""

from pydantic import BaseModel


# --- Bracket Analysis ---


class BracketDetail(BaseModel):
    rate: float
    income_in_bracket: float
    tax_in_bracket: float
    bracket_floor: float
    bracket_ceiling: float | None = None


class BracketRoomResponse(BaseModel):
    current_rate: float
    next_rate: float | None = None
    room_dollars: float


class AccountBalanceDetail(BaseModel):
    name: str
    balance: float


class AccountBalanceSummary(BaseModel):
    tax_deferred: float
    tax_free: float
    taxable: float
    already_taxed: float
    tax_deferred_accounts: list[AccountBalanceDetail] = []
    tax_free_accounts: list[AccountBalanceDetail] = []
    taxable_accounts: list[AccountBalanceDetail] = []


class ACASnapshot(BaseModel):
    magi: float
    fpl_percentage: float
    subsidy_eligible: bool
    monthly_premium: float
    monthly_subsidy: float
    net_monthly_cost: float
    cliff_distance: float
    cliff_warning: bool


class BracketAnalysisResponse(BaseModel):
    filing_status: str
    gross_income: float
    standard_deduction: float
    federal_deduction_method: str = "standard"
    taxable_income: float
    federal_tax: float
    federal_brackets: list[BracketDetail]
    federal_effective_rate: float
    federal_marginal_rate: float
    state_tax: float
    state: str = ""
    state_gross_income: float = 0
    state_rate: float
    state_tax_method: str = "flat_rate"
    state_taxable_income: float = 0
    state_standard_deduction: float = 0
    state_deduction_method: str = "standard"
    state_income_tax: float = 0
    state_payroll_tax: float = 0
    state_effective_rate: float = 0
    state_marginal_rate: float = 0
    fica_tax: float
    total_tax: float
    overall_effective_rate: float
    bracket_room: BracketRoomResponse
    account_balances: AccountBalanceSummary
    aca: ACASnapshot


# --- Withdrawal Plan ---


class WithdrawalYearResponse(BaseModel):
    year: int
    age: float
    spending_need: float = 0.0
    taxes_funded: float = 0.0
    net_spendable: float = 0.0
    from_taxable: float
    from_deferred: float
    from_roth: float
    from_cash: float = 0.0
    roth_conversion: float
    total_income: float
    ordinary_income: float
    capital_gains_income: float
    federal_tax: float
    state_tax: float
    fica_tax: float
    total_tax: float
    effective_rate: float
    after_tax_income: float
    magi: float


class WithdrawalPlanResponse(BaseModel):
    years: list[WithdrawalYearResponse]
    total_tax_paid: float
    average_effective_rate: float
    total_withdrawn: float


# --- Roth Conversion Plan ---


class RothConversionYearResponse(BaseModel):
    year: int
    age: float
    baseline_income: float
    conversion_amount: float
    total_taxable_income: float
    tax_on_conversion: float
    cumulative_converted: float
    magi_after_conversion: float
    bracket_filled_to: float


class RothConversionPlanResponse(BaseModel):
    years: list[RothConversionYearResponse]
    total_converted: float
    total_tax_paid: float
    estimated_tax_saved: float
    target_bracket_rate: float
    conversion_window: str
    # Savings = converted × this assumed future-RMD marginal rate − tax paid.
    # Config-driven (custom_assumptions.tax.assumed_rmd_marginal_rate).
    assumed_rmd_marginal_rate: float = 0.24


# --- ACA Analysis ---


class ACAAnalysisResponse(BaseModel):
    magi: float
    household_size: int
    fpl: float
    fpl_percentage: float
    subsidy_eligible: bool
    estimated_monthly_premium: float
    estimated_monthly_subsidy: float
    estimated_net_monthly_cost: float
    cliff_distance: float
    cliff_warning: bool


# --- Tax Scenario ---


class TaxSituationSnapshot(BaseModel):
    gross_income: float
    taxable_income: float
    total_tax: float
    effective_rate: float


class TaxScenarioSnapshot(BaseModel):
    gross_income: float
    taxable_income: float
    total_tax: float
    effective_rate: float
    roth_conversion: float
    extra_income: float
    extra_deduction: float


class TaxScenarioDelta(BaseModel):
    additional_tax: float
    marginal_rate_on_new_income: float


class TaxScenarioResponse(BaseModel):
    base: TaxSituationSnapshot
    scenario: TaxScenarioSnapshot
    delta: TaxScenarioDelta


class TaxScenarioInput(BaseModel):
    roth_conversion: float = 0
    extra_income: float = 0
    extra_deduction: float = 0


# --- Monte Carlo ---


class PercentileCurvePoint(BaseModel):
    age: float
    p10: float
    p25: float
    p50: float
    p75: float
    p90: float


class MonteCarloResponse(BaseModel):
    success_rate: float
    percentile_10: float
    percentile_25: float
    percentile_50: float
    percentile_75: float
    percentile_90: float
    mean_final_nw: float
    total_runs: int
    percentile_curves: list[PercentileCurvePoint]
    worst_final_nw: float
    best_final_nw: float
    # Model disclosure (additive; frontend tolerates absence)
    assumptions: dict | None = None


# --- SEPP / 72(t) ---


class SEPPMethodResult(BaseModel):
    annual: float
    monthly: float


class SEPPReverseResult(BaseModel):
    target_monthly: float
    required_balance: float  # IRA-A size that yields the target (amortization)


class SEPPResponse(BaseModel):
    balance: float
    age: int
    rate: float
    max_rate: float  # max(5%, 120% of mid-term AFR) — Notice 2022-6
    life_expectancy: float  # Single Life Table divisor used
    amortization: SEPPMethodResult
    rmd_method: SEPPMethodResult  # year-1 value; redetermined annually by design
    annuitization: SEPPMethodResult | None = None  # not implemented in v1
    reverse: SEPPReverseResult | None = None
    assumptions: dict

File: frontend/src/types/tax.ts
// --- Bracket Analysis ---

export interface BracketDetail {
  rate: number;
  income_in_bracket: number;
  tax_in_bracket: number;
  bracket_floor: number;
  bracket_ceiling: number | null;
}

export interface BracketRoomInfo {
  current_rate: number;
  next_rate: number | null;
  room_dollars: number;
}

export interface AccountBalanceDetail {
  name: string;
  balance: number;
  cost_basis?: number | null;
  cost_basis_pct?: number | null;
  basis_source?: string | null;
}

export interface AccountBalanceSummary {
  tax_deferred: number;
  tax_free: number;
  taxable: number;
  already_taxed: number;
  tax_deferred_accounts: AccountBalanceDetail[];
  tax_free_accounts: AccountBalanceDetail[];
  taxable_accounts: AccountBalanceDetail[];
  taxable_cost_basis: number;
  taxable_cost_basis_pct: number;
  taxable_basis_coverage_pct: number;
}

export interface ACASnapshot {
  magi: number;
  fpl_percentage: number;
  subsidy_eligible: boolean;
  monthly_premium: number;
  monthly_subsidy: number;
  net_monthly_cost: number;
  cliff_distance: number;
  cliff_warning: boolean;
}

export interface BracketAnalysis {
  filing_status: string;
  gross_income: number;
  standard_deduction: number;
  federal_deduction_method: string;
  federal_itemized_deduction: number;
  federal_salt_deduction: number;
  federal_mortgage_interest: number;
  taxable_income: number;
  federal_tax: number;
  federal_brackets: BracketDetail[];
  federal_effective_rate: number;
  federal_marginal_rate: number;
  state_tax: number;
  state: string;
  state_gross_income: number;
  state_rate: number;
  state_tax_method: string;
  state_taxable_income: number;
  state_standard_deduction: number;
  state_deduction_method: string;
  state_itemized_before_limit: number;
  state_itemized_limitation: number;
  state_mortgage_interest: number;
  state_income_tax: number;
  state_payroll_tax: number;
  state_effective_rate: number;
  state_marginal_rate: number;
  fica_tax: number;
  total_tax: number;
  overall_effective_rate: number;
  bracket_room: BracketRoomInfo;
  account_balances: AccountBalanceSummary;
  aca: ACASnapshot;
}

// --- Withdrawal Plan ---

export interface WithdrawalYear {
  year: number;
  age: number;
  spending_need: number;
  taxes_funded: number;
  net_spendable: number;
  from_taxable: number;
  from_deferred: number;
  from_roth: number;
  from_cash: number;
  roth_conversion: number;
  total_income: number;
  ordinary_income: number;
  capital_gains_income: number;
  federal_tax: number;
  state_tax: number;
  fica_tax: number;
  total_tax: number;
  effective_rate: number;
  after_tax_income: number;
  magi: number;
}

export interface WithdrawalPlan {
  years: WithdrawalYear[];
  total_tax_paid: number;
  average_effective_rate: number;
  total_withdrawn: number;
}

// --- Roth Conversion Plan ---

export interface RothConversionYear {
  year: number;
  age: number;
  baseline_income: number;
  conversion_amount: number;
  total_taxable_income: number;
  tax_on_conversion: number;
  cumulative_converted: number;
  magi_after_conversion: number;
  bracket_filled_to: number;
}

export interface RothConversionPlan {
  years: RothConversionYear[];
  total_converted: number;
  total_tax_paid: number;
  estimated_tax_saved: number;
  target_bracket_rate: number;
  conversion_window: string;
  /** Assumed future-RMD marginal rate behind estimated_tax_saved. */
  assumed_rmd_marginal_rate: number;
}

// --- SEPP / 72(t) ---

export interface SEPPMethodResult {
  annual: number;
  monthly: number;
}

export interface SEPPReverseResult {
  target_monthly: number;
  /** IRA-A size that yields the target under fixed amortization. */
  required_balance: number;
}

export interface SEPPResponse {
  balance: number;
  age: number;
  rate: number;
  /** max(5%, 120% of mid-term AFR) — Notice 2022-6 */
  max_rate: number;
  life_expectancy: number;
  amortization: SEPPMethodResult;
  /** Year-1 value; redetermined annually by design. */
  rmd_method: SEPPMethodResult;
  annuitization: SEPPMethodResult | null;
  reverse: SEPPReverseResult | null;
  assumptions: Record<string, string>;
}

// --- Tax Scenario ---

export interface TaxScenarioInput {
  roth_conversion?: number;
  extra_income?: number;
  extra_deduction?: number;
}

export interface TaxScenarioResponse {
  base: {
    gross_income: number;
    taxable_income: number;
    total_tax: number;
    effective_rate: number;
  };
  scenario: {
    gross_income: number;
    taxable_income: number;
    total_tax: number;
    effective_rate: number;
    roth_conversion: number;
    extra_income: number;
    extra_deduction: number;
  };
  delta: {
    additional_tax: number;
    marginal_rate_on_new_income: number;
  };
}

// --- Monte Carlo ---

export interface PercentileCurvePoint {
  age: number;
  p10: number;
  p25: number;
  p50: number;
  p75: number;
  p90: number;
}

export interface MonteCarloResult {
  success_rate: number;
  percentile_10: number;
  percentile_25: number;
  percentile_50: number;
  percentile_75: number;
  percentile_90: number;
  mean_final_nw: number;
  total_runs: number;
  percentile_curves: PercentileCurvePoint[];
  worst_final_nw: number;
  best_final_nw: number;
  starting_spendable_assets: number;
  excluded_non_spendable_assets: number;
  assumptions?: Record<string, unknown> | null;
}

export interface RetirementConfidenceAge {
  confidence: number;
  earliest_age: number | null;
  success_rate: number | null;
  prior_age_success_rate: number | null;
}

export interface RetirementAgeAnalysis {
  configured_retirement_age: number | null;
  current_age: number;
  max_tested_age: number;
  runs_per_age: number;
  confidence_ages: RetirementConfidenceAge[];
}
